scipy/optimize/_trustregion_exact.py

```diff
@@ -0,0 +1,424 @@
+                                  **trust_region_options)
+
+
+def estimate_smallest_singular_value(c):
+    """Given upper triangular matrix C estimate the smallest singular
+    value and the correspondent right singular vector in O(n**2) operations.
+
+    Parameters
+    ----------
+    c : ndarray
+        Square upper triangular matrix.
+
+    Returns
+    -------
+    smin : float
+        Estimated smallest singular value of the provided matrix.
+    zmin : ndarray
+        Estimatied right singular vector.
+
+    Notes
+    -----
+    The procedure is based on [1]_ and is done in two steps. First it finds
+    a vector ``e`` with components selected from {+1, -1} such that the
+    solution ``w`` from the system ``C.T w = e`` is as large as possible.
+    Next it estimate ``C v = w``. The smallest singular value is close
+    to ``norm(w)/norm(v)`` and the right singular vector is close
+    to ``v/norm(v)``. Both [1]_ and [2]_  provide some insight about the
+    reasons behind this procedure.
+
+    The estimation will be better more ill-conditioned is the matrix.
+
+    References
+    ----------
+    .. [1] Cline, A. K., Moler, C. B., Stewart, G. W., Wilkinson, J. H. (1979).
+           An estimate for the condition number of a matrix.
+           SIAM Journal on Numerical Analysis, 16(2), 368-375.
+    .. [2] Golub, G. H., Van Loan, C. F. (2013).
+           "Matrix computations". Forth Edition.
+           JHU press. pp. 140-142
+    """
+
+    c = np.atleast_2d(c)
+    m, n = c.shape
+
+    if m != n:
+        raise ValueError("A square triangular matrix should be provided.")
+
+    # A vector `e` with components selected from {+1, -1}
+    # is selected so that the solution `w` to the system
+    # `C.T w = e` is as large as possible. Implementation
+    # based on algorithm 3.5.1, p. 142, from reference [2]
+    # adapted for lower triangular matrix.
+
+    p = np.zeros(n)
+    w = np.empty(n)
+
+    for k in range(n):
+        wp = (1-p[k])/c.T[k, k]
+        wm = (-1-p[k])/c.T[k, k]
+        pp = p[k+1:]+c.T[k+1:, k]*wp
+        pm = p[k+1:]+c.T[k+1:, k]*wm
+
+        if abs(wp)+norm(pp, 1) >= abs(wm)+norm(pm, 1):
+            w[k] = wp
+            p[k+1:] = pp
+        else:
+            w[k] = wm
+            p[k+1:] = pm
+
+    # The system `C v = w` is solved using backward substitution.
+    v = solve_triangular(c, w)
+
+    # Compute vector norms
+    vnorm = norm(v)
+    wnorm = norm(w)
+
+    # Smallest singular value
+    smin = wnorm/vnorm
+
+    # Associated vector
+    zmin = v/vnorm
+
+    return smin, zmin
+
+
+def gershgorin_bounds(H):
+    """
+    Given a square matrix ``M`` compute upper
+    and lower bounds for its eigenvalues (Gregoshgorin Bounds).
+    Defined ref. [1].
+
+    References
+    ----------
+    .. [1] Conn, A. R., Gould, N. I., & Toint, P. L. (2000).
+           "Trust region methods". Siam. pp. 19.
+    """
+
+    D = np.diag(np.diag(H))
```

Comment:
This implementation looks suspicious to me:

1. I suggest to use `np.` functions.
2. `sum` is used without axis specified, which can be a bug.

I suggest the following implementation, which doesn't create additional matrices (please check):
```Python
H_diag = np.diag(H)
H_diag_abs = np.abs(H_diag)
H_row_sums = np.sum(np.abs(H), axis=1)
lb = np.max(H_diag + H_diag_abs - H_row_sums)
ub = np.max(H_diag - H_diag_abs + H_row_sums)
```